### PR TITLE
CSHARP-2739: improve hierarchical fields projection.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Evergreen: https://evergreen.mongodb.com/version/5d7d4e8c5623431a16d885a2
Original PR: https://github.com/mongodb/mongo-csharp-driver/pull/383

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Translators/FindProjectionTranslator.cs [hunk 1]
```diff
@@ -218,12 +218,9 @@ private static IReadOnlyList<IFieldExpression> GetUniqueFieldsByHierarchy(IEnume
                 var referenceGroup = referenceGroups.Dequeue();
                 if (!skippedFields.Contains(referenceGroup.Key))
                 {
-                    var hierarchicalReferenceGroups = referenceGroups
-                        .Where(x =>
-                        {
-                            return x.Key.Count(c => c == '.') != referenceGroup.Key.Count(c => c == '.') && // otherwise, fields are on the same nesting level
-                                   x.Key.StartsWith(referenceGroup.Key);
-                        });
+                    var prefix = referenceGroup.Key + '.';
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/FindProjectionTranslator.cs:221]: NOTE: 
with this line:
```
referenceGroup.Key + '.'
```
checking on the number of `dot`s is not obligatory.

NOTE2: 
Linq queries similar to this one (without a `dot` in one from fields): 
```
Project(p => new {  F = p, a = p.C }, "{ C : { D : '1', E1 : { F : 2 } } }")
```

which have a document itself doesn't generate `Projection` field. See this test: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Tests/Linq/Translators/FindProjectionTranslatorTests.cs#L47
So, they are correct.

File: tests/MongoDB.Driver.Tests/Linq/Translators/FindProjectionTranslatorTests.cs [hunk 2]
```diff
@@ -193,25 +193,25 @@ public void Should_translate_with_a_hierarchical_redundancy_and_a_top_level_proj
         }
 
         [Fact]
-        public void Should_translate_with_a_hierarchical_redundancy_and_when_one_field_starts_with_another()
+        public void Should_translate_with_a_hierarchical_redundancy_and_when_one_field_doesnt_start_with_another()
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Linq/Translators/FindProjectionTranslatorTests.cs:196]: changed the tests order

**Review by rstam — COMMENTED** [review]
I have no idea whether this change is correct or not.

I'm going to remove myself as a reviewer.

**Review by rstam — APPROVED** [review]
I've looked at the change a bit more closely and I am satisfied that the change made is just a different way of accomplishing essentially the same test as before (in a simpler way).

**Pull request closed without merging**